## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [ ]:
# pacotes usados neste notebook
import pandas as pd

<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *amostra (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [ ]:
data = pd.read_csv('/GasPricesinBrazil_2004-2019.csv')

In [ ]:
data

Unnamed: 0;DATA INICIAL;DATA FINAL;REGIÃO;ESTADO;PRODUTO;NÚMERO DE POSTOS PESQUISADOS;UNIDADE DE MEDIDA;PREÇO MÉDIO REVENDA;DESVIO PADRÃO REVENDA;PREÇO MÍNIMO REVENDA;PREÇO MÁXIMO REVENDA;MARGEM MÉDIA REVENDA;COEF DE VARIAÇÃO REVENDA;PREÇO MÉDIO DISTRIBUIÇÃO;DESVIO PADRÃO DISTRIBUIÇÃO;PREÇO MÍNIMO DISTRIBUIÇÃO;PREÇO MÁXIMO DISTRIBUIÇÃO;COEF DE VARIAÇÃO DISTRIBUIÇÃO;MÊS;ANO
0       0;2004-05-09;2004-05-15;CENTRO OESTE;DISTRITO ...                                                                                                                                                                                                                                                                                                                                     
1       1;2004-05-09;2004-05-15;CENTRO OESTE;GOIAS;ETA...                                                                                                                                                                                                                                                                                                                                     
2       2;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...                                                                                                                                                                                                                                                                                                                                     
3       3;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...                                                                                                                                                                                                                                                                                                                                     
4       4;2004-05-09;2004-05-15;NORDESTE;ALAGOAS;ETANO...                                                                                                                                                                                                                                                                                                                                     
...                                                   ...                                                                                                                                                                                                                                                                                                                                     
106818  106818;2019-06-23;2019-06-29;NORDESTE;RIO GRAN...                                                                                                                                                                                                                                                                                                                                     
106819  106819;2019-06-23;2019-06-29;SUL;RIO GRANDE DO...                                                                                                                                                                                                                                                                                                                                     
106820  106820;2019-06-23;2019-06-29;SUL;SANTA CATARIN...                                                                                                                                                                                                                                                                                                                                     
106821  106821;2019-06-23;2019-06-29;SUDESTE;SAO PAULO...                                                                                                                        

O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

In [ ]:
#Carregando o dataset corretamente ==> neste caso, usa o separador ';'
data = pd.read_csv('/content/GasPricesinBrazil_2004-2019.csv', sep=';')

In [ ]:
data

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO  \
0                0   2004-05-09  2004-05-15  CENTRO OESTE   
1                1   2004-05-09  2004-05-15  CENTRO OESTE   
2                2   2004-05-09  2004-05-15  CENTRO OESTE   
3                3   2004-05-09  2004-05-15  CENTRO OESTE   
4                4   2004-05-09  2004-05-15      NORDESTE   
...            ...          ...         ...           ...   
106818      106818   2019-06-23  2019-06-29      NORDESTE   
106819      106819   2019-06-23  2019-06-29           SUL   
106820      106820   2019-06-23  2019-06-29           SUL   
106821      106821   2019-06-23  2019-06-29       SUDESTE   
106822      106822   2019-06-23  2019-06-29      NORDESTE   

                     ESTADO           PRODUTO  NÚMERO DE POSTOS PESQUISADOS  \
0          DISTRITO FEDERAL  ETANOL HIDRATADO                           127   
1                     GOIAS  ETANOL HIDRATADO                           387   
2               MATO GROSSO  ETANOL HIDRATADO                           192   
3        MATO GROSSO DO SUL  ETANOL HIDRATADO                           162   
4                   ALAGOAS  ETANOL HIDRATADO                           103   
...                     ...               ...                           ...   
106818  RIO GRANDE DO NORTE               GNV                             7   
106819    RIO GRANDE DO SUL               GNV                            23   
106820       SANTA CATARINA               GNV                            24   
106821            SAO PAULO               GNV                            52   
106822              SERGIPE               GNV                             4   

       UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  \
0                   R$/l                1.288                  0.016  ...   
1                   R$/l                1.162                  0.114  ...   
2                   R$/l                1.389                  0.097  ...   
3                   R$/l                1.262                  0.070  ...   
4                   R$/l                1.181                  0.078  ...   
...                  ...                  ...                    ...  ...   
106818             R$/m3                3.574                  0.065  ...   
106819             R$/m3                3.401                  0.129  ...   
106820             R$/m3                2.912                  0.190  ...   
106821             R$/m3                3.020                  0.229  ...   
106822             R$/m3                3.697                  0.005  ...   

        PREÇO MÁXIMO REVENDA  MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  \
0                      1.350                 0.463                    0.012   
1                      1.449                 0.399                    0.098   
2                      1.760                 0.419                    0.070   
3                      1.509                 0.432                    0.055   
4                      1.400                  0.24                    0.066   
...                      ...                   ...                      ...   
106818                 3.690                 0.818                    0.018   
106819                 3.789                  0.95                    0.038   
106820                 3.499                 0.914                    0.065   
106821                 3.490                 0.646                    0.076   
106822                 3.699                 1.179                    0.001   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                          0.825                       0.11   
1                          0.763                      0.088   
2                           0.97                      0.095   
3                           0.83                      0.119   
4                          0.941                      0.077   
...                          ...                        ...   
106818     

### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [ ]:
data.head(2)

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE             GOIAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  

[2 rows x 21 columns]

### 1.3 Informações do Dataset e Elementos Chave

### 1.3.1 Informações gerais sobre o Dataset

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 21 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Unnamed: 0                     106823 non-null  int64  
 1   DATA INICIAL                   106823 non-null  object 
 2   DATA FINAL                     106823 non-null  object 
 3   REGIÃO                         106823 non-null  object 
 4   ESTADO                         106823 non-null  object 
 5   PRODUTO                        106823 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 7   UNIDADE DE MEDIDA              106823 non-null  object 
 8   PREÇO MÉDIO REVENDA            106823 non-null  float64
 9   DESVIO PADRÃO REVENDA          106823 non-null  float64
 10  PREÇO MÍNIMO REVENDA           106823 non-null  float64
 11  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 12  MARGEM MÉDIA REVENDA          

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

In [ ]:
type(data)

pandas.core.frame.DataFrame

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

In [ ]:
data.shape

(106823, 21)

In [ ]:
print(f'O DataFrame Possui {data.shape[0]} linhas/observações. E a Kelly vai me dar {data.shape[1]} sentadas')

O DataFrame Possui 106823 linhas/observações. E a Kelly vai me dar 21 sentadas


#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [ ]:
personagens_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70.5, 15.2, 60.1],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [ ]:
personagens_df

nome  idade  peso  eh jedi
0  Luke Skywalker     16  70.5     True
1            Yoda   1000  15.2     True
2       Palpatine     70  60.1    False

In [ ]:
personagens_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   nome     3 non-null      object 
 1   idade    3 non-null      int64  
 2   peso     3 non-null      float64
 3   eh jedi  3 non-null      bool   
dtypes: bool(1), float64(1), int64(1), object(1)
memory usage: 203.0+ bytes


#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O método `DataFrame.columns` retorna uma ";lista" com os **nomes de todas as colunas** do data frame.

In [ ]:
personagens_df.columns

Index(['nome', 'idade', 'peso', 'eh jedi'], dtype='object')

In [ ]:
type(personagens_df.columns)

pandas.core.indexes.base.Index

In [ ]:
list(personagens_df.columns)

['nome', 'idade', 'peso', 'eh jedi']

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

In [ ]:
personagens_df_renomeado = personagens_df.rename(columns={
    'nome': 'Kelly safada',
    'idade': 'Idade'
})

In [ ]:
personagens_df_renomeado

Kelly safada  Idade  peso  eh jedi
0  Luke Skywalker     16  70.5     True
1            Yoda   1000  15.2     True
2       Palpatine     70  60.1    False

Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

In [ ]:
personagens_df.rename(columns={
    'nome': 'Kelly safada',
    'idade': 'Idade'
}
,inplace=True
)

In [ ]:
personagens_df

Kelly safada  Idade  peso  eh jedi
0  Luke Skywalker     16  70.5     True
1            Yoda   1000  15.2     True
2       Palpatine     70  60.1    False

<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

In [ ]:
personagens_df.columns = ['NOME','IDADE','PESO','EH_JEDI']

In [ ]:
personagens_df

NOME  IDADE  PESO  EH_JEDI
0  Luke Skywalker     16  70.5     True
1            Yoda   1000  15.2     True
2       Palpatine     70  60.1    False

### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

In [ ]:
data.head()

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   
2                1.389                  0.097  ...                 1.760   
3                1.262                  0.070  ...                 1.509   
4                1.181                  0.078  ...                 1.400   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   
2                 0.419                    0.070                      0.97   
3                 0.432                    0.055                      0.83   
4                  0.24                    0.066                     0.941   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  
2                     1.161                         0.098   5  2004  
3                   1.22242                         0.143   5  2004  
4                    1.0317                         0.082   5  2004  

[5 rows x 21 columns]

In [ ]:
data['ESTADO']

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [ ]:
data.ESTADO

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [ ]:
type(data.ESTADO)

pandas.core.series.Series

In [ ]:
data.iloc[2]

Unnamed: 0                                      2
DATA INICIAL                           2004-05-09
DATA FINAL                             2004-05-15
REGIÃO                               CENTRO OESTE
ESTADO                                MATO GROSSO
PRODUTO                          ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS                  192
UNIDADE DE MEDIDA                            R$/l
PREÇO MÉDIO REVENDA                         1.389
DESVIO PADRÃO REVENDA                       0.097
PREÇO MÍNIMO REVENDA                         1.18
PREÇO MÁXIMO REVENDA                         1.76
MARGEM MÉDIA REVENDA                        0.419
COEF DE VARIAÇÃO REVENDA                     0.07
PREÇO MÉDIO DISTRIBUIÇÃO                     0.97
DESVIO PADRÃO DISTRIBUIÇÃO                  0.095
PREÇO MÍNIMO DISTRIBUIÇÃO                  0.5614
PREÇO MÁXIMO DISTRIBUIÇÃO                   1.161
COEF DE VARIAÇÃO DISTRIBUIÇÃO               0.098
MÊS                                             5


In [ ]:
pd.Series([5.5, 6.0, 9.5], index=['prova 1','prova 2','projeto'], name='Notas do Luke Skywalker')

prova 1    5.5
prova 2    6.0
projeto    9.5
Name: Notas do Luke Skywalker, dtype: float64

#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

#### 1.3.4.1 Atribuindo constantes

In [ ]:
data

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO  \
0                0   2004-05-09  2004-05-15  CENTRO OESTE   
1                1   2004-05-09  2004-05-15  CENTRO OESTE   
2                2   2004-05-09  2004-05-15  CENTRO OESTE   
3                3   2004-05-09  2004-05-15  CENTRO OESTE   
4                4   2004-05-09  2004-05-15      NORDESTE   
...            ...          ...         ...           ...   
106818      106818   2019-06-23  2019-06-29      NORDESTE   
106819      106819   2019-06-23  2019-06-29           SUL   
106820      106820   2019-06-23  2019-06-29           SUL   
106821      106821   2019-06-23  2019-06-29       SUDESTE   
106822      106822   2019-06-23  2019-06-29      NORDESTE   

                     ESTADO           PRODUTO  NÚMERO DE POSTOS PESQUISADOS  \
0          DISTRITO FEDERAL  ETANOL HIDRATADO                           127   
1                     GOIAS  ETANOL HIDRATADO                           387   
2               MATO GROSSO  ETANOL HIDRATADO                           192   
3        MATO GROSSO DO SUL  ETANOL HIDRATADO                           162   
4                   ALAGOAS  ETANOL HIDRATADO                           103   
...                     ...               ...                           ...   
106818  RIO GRANDE DO NORTE               GNV                             7   
106819    RIO GRANDE DO SUL               GNV                            23   
106820       SANTA CATARINA               GNV                            24   
106821            SAO PAULO               GNV                            52   
106822              SERGIPE               GNV                             4   

       UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  \
0                   R$/l                1.288                  0.016  ...   
1                   R$/l                1.162                  0.114  ...   
2                   R$/l                1.389                  0.097  ...   
3                   R$/l                1.262                  0.070  ...   
4                   R$/l                1.181                  0.078  ...   
...                  ...                  ...                    ...  ...   
106818             R$/m3                3.574                  0.065  ...   
106819             R$/m3                3.401                  0.129  ...   
106820             R$/m3                2.912                  0.190  ...   
106821             R$/m3                3.020                  0.229  ...   
106822             R$/m3                3.697                  0.005  ...   

        PREÇO MÁXIMO REVENDA  MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  \
0                      1.350                 0.463                    0.012   
1                      1.449                 0.399                    0.098   
2                      1.760                 0.419                    0.070   
3                      1.509                 0.432                    0.055   
4                      1.400                  0.24                    0.066   
...                      ...                   ...                      ...   
106818                 3.690                 0.818                    0.018   
106819                 3.789                  0.95                    0.038   
106820                 3.499                 0.914                    0.065   
106821                 3.490                 0.646                    0.076   
106822                 3.699                 1.179                    0.001   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                          0.825                       0.11   
1                          0.763                      0.088   
2                           0.97                      0.095   
3                           0.83                      0.119   
4                          0.941                      0.077   
...                          ...                        ...   
106818     

In [ ]:
produto_view = data['PRODUTO']
produto_view

0         Combustível
1         Combustível
2         Combustível
3         Combustível
4         Combustível
             ...     
106818    Combustível
106819    Combustível
106820    Combustível
106821    Combustível
106822    Combustível
Name: PRODUTO, Length: 106823, dtype: object

In [ ]:
produto_copy_bkp = data['PRODUTO'].copy()
produto_copy_bkp

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [ ]:
data['PRODUTO']='Combustível'

In [ ]:
data.head()

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

       PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  Combustível                           127              R$/l   
1  Combustível                           387              R$/l   
2  Combustível                           192              R$/l   
3  Combustível                           162              R$/l   
4  Combustível                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   
2                1.389                  0.097  ...                 1.760   
3                1.262                  0.070  ...                 1.509   
4                1.181                  0.078  ...                 1.400   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   
2                 0.419                    0.070                      0.97   
3                 0.432                    0.055                      0.83   
4                  0.24                    0.066                     0.941   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  
2                     1.161                         0.098   5  2004  
3                   1.22242                         0.143   5  2004  
4                    1.0317                         0.082   5  2004  

[5 rows x 21 columns]

In [ ]:
produto_view

0         Combustível
1         Combustível
2         Combustível
3         Combustível
4         Combustível
             ...     
106818    Combustível
106819    Combustível
106820    Combustível
106821    Combustível
106822    Combustível
Name: PRODUTO, Length: 106823, dtype: object

#### 1.3.4.2 Atribuindo listas ou series

In [ ]:
nrows, ncols = data.shape
nrows

106823

In [ ]:
novos_produtos = [f'Produto {i}' for i in range(nrows)]
novos_produtos

['Produto 0',
 'Produto 1',
 'Produto 2',
 'Produto 3',
 'Produto 4',
 'Produto 5',
 'Produto 6',
 'Produto 7',
 'Produto 8',
 'Produto 9',
 'Produto 10',
 'Produto 11',
 'Produto 12',
 'Produto 13',
 'Produto 14',
 'Produto 15',
 'Produto 16',
 'Produto 17',
 'Produto 18',
 'Produto 19',
 'Produto 20',
 'Produto 21',
 'Produto 22',
 'Produto 23',
 'Produto 24',
 'Produto 25',
 'Produto 26',
 'Produto 27',
 'Produto 28',
 'Produto 29',
 'Produto 30',
 'Produto 31',
 'Produto 32',
 'Produto 33',
 'Produto 34',
 'Produto 35',
 'Produto 36',
 'Produto 37',
 'Produto 38',
 'Produto 39',
 'Produto 40',
 'Produto 41',
 'Produto 42',
 'Produto 43',
 'Produto 44',
 'Produto 45',
 'Produto 46',
 'Produto 47',
 'Produto 48',
 'Produto 49',
 'Produto 50',
 'Produto 51',
 'Produto 52',
 'Produto 53',
 'Produto 54',
 'Produto 55',
 'Produto 56',
 'Produto 57',
 'Produto 58',
 'Produto 59',
 'Produto 60',
 'Produto 61',
 'Produto 62',
 'Produto 63',
 'Produto 64',
 'Produto 65',
 'Produto 66',
 'Pro

In [ ]:
data['PRODUTO'] = novos_produtos

In [ ]:
data

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO  \
0                0   2004-05-09  2004-05-15  CENTRO OESTE   
1                1   2004-05-09  2004-05-15  CENTRO OESTE   
2                2   2004-05-09  2004-05-15  CENTRO OESTE   
3                3   2004-05-09  2004-05-15  CENTRO OESTE   
4                4   2004-05-09  2004-05-15      NORDESTE   
...            ...          ...         ...           ...   
106818      106818   2019-06-23  2019-06-29      NORDESTE   
106819      106819   2019-06-23  2019-06-29           SUL   
106820      106820   2019-06-23  2019-06-29           SUL   
106821      106821   2019-06-23  2019-06-29       SUDESTE   
106822      106822   2019-06-23  2019-06-29      NORDESTE   

                     ESTADO         PRODUTO  NÚMERO DE POSTOS PESQUISADOS  \
0          DISTRITO FEDERAL       Produto 0                           127   
1                     GOIAS       Produto 1                           387   
2               MATO GROSSO       Produto 2                           192   
3        MATO GROSSO DO SUL       Produto 3                           162   
4                   ALAGOAS       Produto 4                           103   
...                     ...             ...                           ...   
106818  RIO GRANDE DO NORTE  Produto 106818                             7   
106819    RIO GRANDE DO SUL  Produto 106819                            23   
106820       SANTA CATARINA  Produto 106820                            24   
106821            SAO PAULO  Produto 106821                            52   
106822              SERGIPE  Produto 106822                             4   

       UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  \
0                   R$/l                1.288                  0.016  ...   
1                   R$/l                1.162                  0.114  ...   
2                   R$/l                1.389                  0.097  ...   
3                   R$/l                1.262                  0.070  ...   
4                   R$/l                1.181                  0.078  ...   
...                  ...                  ...                    ...  ...   
106818             R$/m3                3.574                  0.065  ...   
106819             R$/m3                3.401                  0.129  ...   
106820             R$/m3                2.912                  0.190  ...   
106821             R$/m3                3.020                  0.229  ...   
106822             R$/m3                3.697                  0.005  ...   

        PREÇO MÁXIMO REVENDA  MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  \
0                      1.350                 0.463                    0.012   
1                      1.449                 0.399                    0.098   
2                      1.760                 0.419                    0.070   
3                      1.509                 0.432                    0.055   
4                      1.400                  0.24                    0.066   
...                      ...                   ...                      ...   
106818                 3.690                 0.818                    0.018   
106819                 3.789                  0.95                    0.038   
106820                 3.499                 0.914                    0.065   
106821                 3.490                 0.646                    0.076   
106822                 3.699                 1.179                    0.001   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                          0.825                       0.11   
1                          0.763                      0.088   
2                           0.97                      0.095   
3                           0.83                      0.119   
4                          0.941                      0.077   
...                          ...                        ...   
106818                     2.756   

In [ ]:
print(produto_view)
print('/n')
print(produto_copy_bkp)

#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

In [ ]:
data['coluna sem nocao'] = 'DEFAULT'
data

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0               0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1               1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2               2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3               3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4               4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
...           ...          ...         ...           ...                 ...   
45116       45116   2005-06-19  2005-06-25  CENTRO OESTE  MATO GROSSO DO SUL   
45117       45117   2005-06-19  2005-06-25      NORDESTE             ALAGOAS   
45118       45118   2005-06-19  2005-06-25      NORDESTE               BAHIA   
45119       45119   2005-06-19  2005-06-25      NORDESTE               CEARA   
45120       45120   2005-06-19  2005-06-25      NORDESTE            MARANHAO   

                PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0      ETANOL HIDRATADO                           127              R$/l   
1      ETANOL HIDRATADO                           387              R$/l   
2      ETANOL HIDRATADO                           192              R$/l   
3      ETANOL HIDRATADO                           162              R$/l   
4      ETANOL HIDRATADO                           103              R$/l   
...                 ...                           ...               ...   
45116       ÓLEO DIESEL                           149              R$/l   
45117       ÓLEO DIESEL                           100              R$/l   
45118       ÓLEO DIESEL                           427              R$/l   
45119       ÓLEO DIESEL                           298              R$/l   
45120       ÓLEO DIESEL                           188              R$/l   

       PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  MARGEM MÉDIA REVENDA  \
0                    1.288                  0.016  ...                 0.463   
1                    1.162                  0.114  ...                 0.399   
2                    1.389                  0.097  ...                 0.419   
3                    1.262                  0.070  ...                 0.432   
4                    1.181                  0.078  ...                  0.24   
...                    ...                    ...  ...                   ...   
45116                1.856                  0.056  ...                 0.179   
45117                1.668                  0.034  ...                 0.169   
45118                1.638                  0.084  ...                 0.154   
45119                1.676                  0.072  ...                 0.145   
45120                1.662                  0.056  ...                 0.153   

       COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                         0.012                    0.825   
1                         0.098                    0.763   
2                         0.070                     0.97   
3                         0.055                     0.83   
4                         0.066                    0.941   
...                         ...                      ...   
45116                     0.030                    1.677   
45117                     0.020                    1.499   
45118                     0.051                    1.484   
45119                     0.043                    1.531   
45120                     0.034                    1.509   

       DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                            0.11                    0.4201   
1                           0.088                    0.5013   
2                           0.095                    0.5614   
3                           0.119                    0.5991   
4                           0.077                    0.7441   
...                           ...                       ...   
45116                 

In [ ]:
data['coluna a partir de lista'] = range(data.shape[0])
data.head(2)

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE             GOIAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  COEF DE VARIAÇÃO REVENDA  \
0                1.288                  0.016  ...                     0.012   
1                1.162                  0.114  ...                     0.098   

   PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                     0.825                       0.11   
1                     0.763                      0.088   

   PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                     0.4201                    0.9666   
1                     0.5013                      1.05   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
0                         0.133   5  2004          DEFAULT   
1                         0.115   5  2004          DEFAULT   

   coluna a partir de lista  
0                         0  
1                         1  

[2 rows x 23 columns]

In [ ]:
data['PREÇO MÉDIO REVENDA (dólares)'] = data['PREÇO MÉDIO REVENDA'] / 6.0

<br/>

Outro exemplo:

In [ ]:
data.head(2)

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE             GOIAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                1.288                  0.016  ...                     0.825   
1                1.162                  0.114  ...                     0.763   

   DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                        0.11                    0.4201   
1                       0.088                    0.5013   

   PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  \
0                     0.9666                         0.133   5  2004   
1                       1.05                         0.115   5  2004   

  coluna sem nocao coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
0          DEFAULT                        0                       0.214667  
1          DEFAULT                        1                       0.193667  

[2 rows x 24 columns]

**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

In [ ]:
data.index

RangeIndex(start=0, stop=45121, step=1)

Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [7]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [ ]:
pesquisa_de_satisfacao.head()

In [ ]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])


In [ ]:
pesquisa_de_satisfacao.head()

bom  ruim  pessimo
XboxOne        50   131       30
Playstation4   21     2       20
Switch        100    30        1

### 1.4 Selecionando uma ou mais amostras (Indexação)

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma amostra/linha:**

In [ ]:
data.iloc[1]

Unnamed: 0                                      1
DATA INICIAL                           2004-05-09
DATA FINAL                             2004-05-15
REGIÃO                               CENTRO OESTE
ESTADO                                      GOIAS
PRODUTO                          ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS                  387
UNIDADE DE MEDIDA                            R$/l
PREÇO MÉDIO REVENDA                         1.162
DESVIO PADRÃO REVENDA                       0.114
PREÇO MÍNIMO REVENDA                         0.89
PREÇO MÁXIMO REVENDA                        1.449
MARGEM MÉDIA REVENDA                        0.399
COEF DE VARIAÇÃO REVENDA                    0.098
PREÇO MÉDIO DISTRIBUIÇÃO                    0.763
DESVIO PADRÃO DISTRIBUIÇÃO                  0.088
PREÇO MÍNIMO DISTRIBUIÇÃO                  0.5013
PREÇO MÁXIMO DISTRIBUIÇÃO                    1.05
COEF DE VARIAÇÃO DISTRIBUIÇÃO               0.115
MÊS                                             5


In [ ]:
data.iloc[:6]

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
5           5   2004-05-09  2004-05-15      NORDESTE               BAHIA   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   
5  ETANOL HIDRATADO                           408              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   
2                1.389                  0.097  ...                 1.760   
3                1.262                  0.070  ...                 1.509   
4                1.181                  0.078  ...                 1.400   
5                1.383                  0.132  ...                 2.050   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   
2                 0.419                    0.070                      0.97   
3                 0.432                    0.055                      0.83   
4                  0.24                    0.066                     0.941   
5                 0.426                    0.095                     0.957   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   
5                      0.128                    0.5686   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  
2                     1.161                         0.098   5  2004  
3                   1.22242                         0.143   5  2004  
4                    1.0317                         0.082   5  2004  
5                      1.35                         0.134   5  2004  

[6 rows x 21 columns]

**Selecionando múltiplias amostras/linhas:**

In [ ]:
data.iloc[[1,5,10,15]]

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO    ESTADO  \
1            1   2004-05-09  2004-05-15  CENTRO OESTE     GOIAS   
5            5   2004-05-09  2004-05-15      NORDESTE     BAHIA   
10          10   2004-05-09  2004-05-15      NORDESTE     PIAUI   
15          15   2004-05-09  2004-05-15         NORTE  AMAZONAS   

             PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
1   ETANOL HIDRATADO                           387              R$/l   
5   ETANOL HIDRATADO                           408              R$/l   
10  ETANOL HIDRATADO                            89              R$/l   
15  ETANOL HIDRATADO                            53              R$/l   

    PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
1                 1.162                  0.114  ...                 1.449   
5                 1.383                  0.132  ...                 2.050   
10                1.573                  0.134  ...                 1.780   
15                1.578                  0.123  ...                 1.850   

    MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
1                  0.399                    0.098                     0.763   
5                  0.426                    0.095                     0.957   
10                 0.494                    0.085                     1.079   
15                 0.441                    0.078                     1.137   

   DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
1                       0.088                    0.5013   
5                       0.128                    0.5686   
10                      0.078                    0.8857   
15                      0.068                    0.9515   

   PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
1                       1.05                         0.115   5  2004  
5                       1.35                         0.134   5  2004  
10                    1.2499                         0.072   5  2004  
15                    1.2109                          0.06   5  2004  

[4 rows x 21 columns]

In [ ]:
data.iloc[1,4]

'GOIAS'

In [9]:
pesquisa_de_satisfacao.loc[['XboxOne', 'Switch'],['ruim','bom']]

ruim  bom
XboxOne   131   50
Switch     30  100

In [ ]:
# e assim por diante!

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

### 1.5 Selecionando um ou mais atributos (colunas)

Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

### 1.6 Removendo um Atributo (Coluna) do Data Frame

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br/>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

Para **filtrarmos** os registros de postos do estado de São Paulo:

O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

##### **==> Alternativa 2: Utilizando o método `query`**

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análise feita para os postos de São Paulo.

#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/amostras do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro e o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos)

In [ ]:
# Não funciona
# data.query('ESTADO=="RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registro cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, de fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

<br/>

Alternativamente:

#### **Selecionando registros dos anos de 2008, 2010 e 2012**

**ALTERNATIVA 1**

**ALTERNATIVA 2**

### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

<h2>2. Preparação dos dados</h2>
<hr/>

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

#### **Dados Numéricos**

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

**Acessando apenas algumas estatísticas**

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

#### Qual é o menor preço mínimo de revenda?

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

#### Quais são os estados considerados?

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='./imagens/apply_map_applymap.png' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [ ]:
df

`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='./imagens/pandas_axis.jpg' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

<img src='./imagens/apply_axis_1.png' width=250/>

<img src='./imagens/apply_axis_0.png' width=250/>

##### Usando `lambda` functions

<img src='./imagens/apply_axis_1_mean.png' width=350/>

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

<h2>5. Agrupamento</h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

<br/>

Também podemos ter agrupamentos por mais de um atributo.

<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [ ]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])
df

<h2>6. Ordenação</h2>
<hr/>

In [ ]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

`.sort_values()`: ordena valores ao longo de um eixo.

Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

<br/>

Podemos ordenar a partir de **mais de uma coluna**:

Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<br/>

Note que o dataframe original **não foi alterado** após a ordenação.

Para alterá-lo, use o argumento `inplace=True`:

<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`.

Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

#### **Estatísticas Descritivas**

### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning